bigtable_input.py

```diff
@@ -0,0 +1,135 @@
+"""Read Minigo game examples from a Bigtable.
+"""
+
+import logging
+import os
+import struct
+from google.cloud import bigtable
+from google.cloud.bigtable import row_filters
+import tensorflow as tf
+
+
+_project_name = os.environ['PROJECT']
+_instance_name = os.environ['CBT_INSTANCE']
+_table_name = os.environ['CBT_TABLE']
+
+_bt_table = bigtable.Client(
+    _project_name).instance(
+        _instance_name).table(
+            _table_name)
+
+_tf_table = tf.contrib.cloud.BigtableClient(
+    _project_name,
+    _instance_name).table(
+        _table_name)
+
+
+def get_latest_game_number():
+    """Return the number of the next game to be stored.
+
+    The state of the table is stored in `metadata:game_counter`
+    in the `table_state` row only.
+    """
+    counter = b'game_counter'
+    table_state = _bt_table.read_row(
+        'table_state',
+        filter_=row_filters.ColumnRangeFilter('metadata', counter, counter))
+    value = table_state.cells['metadata'][counter][0].value
+    return struct.unpack('>q', value)[0]
+
+
+def get_game_range_row_names(game_begin, game_end):
+    """Get the row range containing the given games.
+
+    Sample row name:
+      g_0000000001_m001
+
+    To capture the range of all moves in the two given games,
+    the end row will need to go up to g_00..(N+1).
+
+    Args:
+      game_begin:  an integer of the beginning game number.
+      game_end:  an integer of the ending game number, inclusive.
+
+    Returns:
+      The two string row numbers to pass to Bigtable as the row range.
+    """
+    row_fmt = 'g_{:0>10}_'
+    return row_fmt.format(game_begin), row_fmt.format(game_end + 1)
+
+
+def get_unparsed_moves_from_last_n_games(n,
+                                         moves=2**21,
+                                         column_family='tfexample',
+                                         column='example'):
+    """Get a dataset of serialized TFExamples from the last N games.
+
+    Args:
+      n:  an integer indicating how many past games should be sourced.
+      moves:  an integer indicating how many moves should be sampled
+        from those N games.
+      column_family:  name of the column family containing move examples.
+      column:  name of the column containing move examples.
+
+    Returns:
+      A dataset containing no more than `moves` examples, sampled
+        randomly from the last `n` games in the table.
+    """
+    latest_game = int(get_latest_game_number())
+    logging.info('Latest game: %s', latest_game)
```

Comment:
This should never be invoked from GTP so it's fine to maybe fine to log to stdout but should we keep this uniform and log with util.dbg?

Reply:
No problem with using `utils.dbg` throughout. Changed.